### PR TITLE
Skip loading SSH keys that already exist in the ssh-agent. Fixes #773

### LINKED ISSUE #773: fin create, start, reset, update all keep asking for ssh passphrase
**Description**

I use ssh key passphrases.  When I do, I am continuously prompted to enter a passphrase when I create, start, reset or update fin.

**Steps to reproduce the issue:**

1. Create ssh key with passphrase
2. Use git repo with ssh key
3. Attempt one of the following commands:
 - `fin create`
 - `fin start`
 - `fin reset`
 - `fin update`

**Describe the results you received:**

```
Enter passphrase for /root/.ssh/id_rsa:
```

**Describe the results you expected:**

ssh key added from id_rsa either once at the beginning of the process or making it so that when I add the ssh key globally, it will continue to reuse

See related: https://github.com/docker/machine/issues/4299

**Output of `fin config`:**

<details>
  <summary>fin config output</summary>

```
COMPOSE_PROJECT_NAME_SAFE: lovin
COMPOSE_FILE:
/Users/username/.docksal/stacks/overrides-osxfs.yml
/Users/username/.docksal/stacks/volumes-bind.yml
/Users/username/.docksal/stacks/stack-default.yml
/Users/username/Sites/project/.docksal/docksal.yml
ENV_FILE:
/Users/username/Sites/project/.docksal/docksal.env

PROJECT_ROOT: /Users/username/Sites/project
DOCROOT: docroot
VIRTUAL_HOST: project.docksal
VIRTUAL_HOST_ALIASES: *.project.docksal
IP: 192.168.64.100
MYSQL: 192.168.64.100:32769

Docker Compose configuration
---------------------
services:
  cli:
    dns:
    - 192.168.64.100
    - 8.8.8.8
    environment:
      DOCROOT: docroot
      HOST_GID: '20'
      HOST_UID: '501'
      SECRET_ACAPI_EMAIL: null
      SECRET_ACAPI_KEY: null
      SECRET_BLACKFIRE_CLIENT_ID: null
      SECRET_BLACKFIRE_CLIENT_TOKEN: null
      SECRET_PLATFORMSH_CLI_TOKEN: null
      SECRET_SSH_PRIVATE_KEY: null
      SECRET_TERMINUS_TOKEN: null
      VIRTUAL_HOST: project.docksal
      XDEBUG_CONFIG: remote_connect_back=0 remote_host=192.168.64.1
      XDEBUG_ENABLED: '0'
    hostname: cli
    image: docksal/cli:2.0-php5.6
    volumes:
    - docksal_ssh_agent:/.ssh-agent:ro
    - cli_home:/home/docker:rw
    - /Users/username/Sites/project:/var/www:rw,cached
  db:
    dns:
    - 192.168.64.100
    - 8.8.8.8
    environment:
      MYSQL_ALLOW_EMPTY_PASSWORD: null
      MYSQL_DATABASE: default
      MYSQL_INITDB_SKIP_TZINFO: null
      MYSQL_ONETIME_PASSWORD: null
      MYSQL_PASSWORD: user
      MYSQL_RANDOM_ROOT_PASSWORD: null
      MYSQL_ROOT_PASSWORD: root
      MYSQL_USER: user
    hostname: db
    image: docksal/db:1.1-mysql-5.6
    ports:
    - 0:3306/tcp
    volumes:
    - db_data:/var/lib/mysql:rw
    - project_root:/var/www:ro,nocopy
  web:
    depends_on:
      cli:
        condition: service_started
    dns:
    - 192.168.64.100
    - 8.8.8.8
    environment:
      APACHE_BASIC_AUTH_PASS: null
      APACHE_BASIC_AUTH_USER: null
      APACHE_DOCUMENTROOT: /var/www/docroot
    hostname: web
    image: docksal/web:2.1-apache2.4
    labels:
      io.docksal.cert-name: none
      io.docksal.project-root: /Users/teknorah/Sites/project
      io.docksal.virtual-host: project.docksal,*.project.docksal,project.docksal.*
    volumes:
    - project_root:/var/www:ro,nocopy
version: '2.1'
volumes:
  cli_home: {}
  db_data: {}
  docksal_ssh_agent:
    external: true
    name: docksal_ssh_agent
  project_root:
    driver: local
    driver_opts:
      device: /Users/username/Sites/project
      o: bind
      type: none

---------------------

```

</details>

<br>

**Output of `fin sysinfo`:**

<details>
  <summary>fin sysinfo output</summary>

  ```
  ███  OS
Darwin Mac OS X 10.13.6
Darwin My-MacBook-Pro.local 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64

███  ENVIRONMENT
MODE : Docker for Mac

███  FIN
fin version: 1.69.1

███  DOCKER COMPOSE
EXPECTED VERSION: 1.21.1
docker-compose version 1.22.0, build f46880f
docker-py version: 3.4.1
CPython version: 3.6.4
OpenSSL version: OpenSSL 1.0.2o  27 Mar 2018

███  DOCKER
EXPECTED VERSION: 18.03.1-ce

Client:
 Version:           18.06.0-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        0ffa825
 Built:             Wed Jul 18 19:05:26 2018
 OS/Arch:           darwin/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:13:46 2018
  OS/Arch:          linux/amd64
  Experimental:     true

███  DOCKSAL: PROJECTS
project             STATUS              virtual host                                    project root
lovin               Up 4 minutes        project.docksal,*.project.docksal,project.docksal.*   /Users/username/Sites/project

███  DOCKSAL: VIRTUAL HOSTS
http://*.project.docksal
http://project.docksal.*
http://project.docksal

███  DOCKER: RUNNING CONTAINERS
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS                   PORTS                                                    NAMES
daa2029b06ba        docksal/web:2.1-apache2.4   "httpd-foreground"       4 minutes ago       Up 4 minutes             80/tcp, 443/tcp                                          project_web_1
234e9ff99a4b        docksal/cli:2.0-php5.6      "/opt/startup.sh sup…"   5 minutes ago       Up 4 minutes (healthy)   22/tcp, 3000/tcp, 9000/tcp                               project_cli_1
ddc57967277b        docksal/db:1.1-mysql-5.6    "/entrypoint.sh mysq…"   5 minutes ago       Up 5 minutes             0.0.0.0:32769->3306/tcp                                  project_db_1
173b9632106d        docksal/ssh-agent:1.0       "/run.sh ssh-agent"      6 minutes ago       Up 6 minutes                                                                      docksal-ssh-agent
74d460b2e47f        docksal/dns:1.0             "/opt/entrypoint.sh …"   6 minutes ago       Up 6 minutes             192.168.64.100:53->53/udp                                docksal-dns
c3088fc30a23        docksal/vhost-proxy:1.3     "docker-entrypoint.s…"   6 minutes ago       Up 6 minutes             192.168.64.100:80->80/tcp, 192.168.64.100:443->443/tcp   docksal-vhost-proxy

███  DOCKER: NETWORKS
NETWORK ID          NAME                 DRIVER              SCOPE
2b609aeb6f80        _default             bridge              local
bcef84044366        bridge               bridge              local
692c8eb87338        host                 host                local
a1506e642239        project_default        bridge              local
856252e9dcde        myandroidy_default   bridge              local
a1d056e752bf        none                 null                local
840400874b17        wptest_default       bridge              local

███  HDD Usage
Filesystem      Size   Used  Avail Capacity iused      ifree %iused  Mounted on
/dev/disk1     697Gi   60Gi  637Gi     9%  772197 4294195082    0%   /
devfs          182Ki  182Ki    0Bi   100%     630          0  100%   /dev
map -hosts       0Bi    0Bi    0Bi   100%       0          0  100%   /net
map auto_home    0Bi    0Bi    0Bi   100%       0          0  100%   /home
  ```

</details>

### PR DESCRIPTION
**Pull request by lmakarov**
Fixes #773

### CODE REVIEW

**Comment by lmakarov** [comment]
@achekulaev this is ready for review.

File: bin/fin [hunk 1]
```diff
@@ -3085,53 +3119,59 @@ ssh_add ()
...
+	docker cp "$ssh_path" docksal-ssh-agent:/
 
-	# Only do this when given an argument (key name) or nothing, but not for options (e.g., -l)
-	if [[ ! "${1}" =~ ^- ]]; then
-		# Copy all keys from the host to the ssh-agent container
-		docker cp "$ssh_path" docksal-ssh-agent:/
+	# If we have a key name as the argument, add it to the list of keys to load
+	if [[ "${1}" != "" ]]; then
+		if [[ -f "${ssh_path}/${1}" ]]; then
+			local SECRET_SSH_KEY_${1^^}=${1}
```
Comment (sean-e-dietrich) [inline bin/fin]: Getting error message

```
fin: line 3141: SECRET_SSH_KEY_${1^^}=${1}: bad substitution
```

Comment (sean-e-dietrich) [reply bin/fin]: Maybe change to something like 

```
                        local ssh_key_name=$(echo $1 | tr '[:lower:]' '[:upper:]')
                        local SECRET_SSH_KEY_${ssh_key_name}=${1}
```

**Comment by lmakarov** [comment]
@sean-e-dietrich what file names do you mean?

**Comment by sean-e-dietrich** [comment]
@lmakarov I meant the key name which ends up usually being the file name.

**Comment by lmakarov** [comment]
@sean-e-dietrich are you referring to these?

```
SECRET_SSH_KEY_ID_RSA=id_rsa
SECRET_SSH_KEY_ID_DSA=id_dsa
SECRET_SSH_KEY_ID_ECDSA=id_ecdsa
```

These would normally be added by default when you run `ssh-add` without arguments. I removed the code handling that case and instead explicitly listed the same default keys as secrets to mimic that behavior and simplify the code. I'm not sure these really belong in `$HOME/.docksal/docksal.env`. I'd rather keep it the way it used to be from the user perspective without any visible config changes.

**Comment by sean-e-dietrich** [comment]
No, I'm talking about the case where someone types

```
fin ssh-add new-key
```

It then adds it to the ssh agent. At that point if someone has to run

```
fin reset system ssh-agent
```

Those keys are no longer apart of the ssh-agent UNLESS they are added to the `$HOME/.docksal/docksal.env`

So my suggestion was that if someone types `fin ssh-add new-key`

We add the following to `docksal.env` 

```
SECRET_SSH_KEY_new_key="new-key"
```

**Comment by lmakarov** [comment]
@sean-e-dietrich ok, I see your point.

It's a good enhancement, but not related to the current PR/issue. We should have a separate PR for that.

This may actually be part of a much larger refactoring of the SSH key management in Docksal that was started a while ago (see https://github.com/docksal/docksal/pull/366).

There are two cases:
  - adding a key temporarily (while the ssh-agent is running, default)
  - adding it permanently.

For the second case, I see it more like `fin ssh-add new-key --permanent`.

File: bin/fin (hunk 1 above)
Comment (lmakarov) [reply bin/fin]: I think your bash is not actually bash, by zsh.

I get the same with `zsh`:

```
$ key=server_rsa
$ eval SECRET_SSH_KEY_${key^^}=${key}
zsh: bad substitution
```

With `bash`:

```
bash-4.4$ key=server_rsa
bash-4.4$ eval SECRET_SSH_KEY_${key^^}=${key}
bash-4.4$ echo $SECRET_SSH_KEY_SERVER_RSA
server_rsa
```

File: bin/fin [hunk 2]
```diff
@@ -3066,6 +3066,40 @@ system_stop ()
...
+	fi
+}
+
+# Helper function to check whether an key is already loaded in the ssh-agent
+# This is useful for keys with a passphrase, which otherwise would require the user to re-enter it.
+# @param $1 host path to the ssh key
+_ssh_key_exists ()
+{
+	# Get fingerprints for keys already loaded in the agent
+	# This function may be called multiple times, so we should cache this value
```
Comment (achekulaev) [inline bin/fin]: Soo.. why didn't you cache it then? :)

File: bin/fin [hunk 3]
```diff
@@ -3066,6 +3066,40 @@ system_stop ()
...
+	if [[ "$existing_fingerprints" == "" ]]; then
+		existing_fingerprints=$(_ssh_add -l | awk '{print $2}')
+		export existing_fingerprints
+	fi
+
+	new_fingerprint=$(ssh-keygen -lf ${1} | awk '{print $2}')
```
Comment (achekulaev) [inline bin/fin]: Wouldn't just checking by file name be enough?

Comment (lmakarov) [reply bin/fin]: It would, unless the ssh key is regenerated/replaced (same name, different key).

Comment (achekulaev) [reply bin/fin]: I'm just worried about the time it takes to execute all this. How long does it take for say 3 keys?

**Comment by achekulaev** [comment]
Approved. Do you want to add a test for this new functionality?

**Comment by lmakarov** [comment]
Yeah, there should be a test case for this. I'll work on it.

**Pull request merged**